syntax = "proto2";
package pbtree;
option java_outer_classname = "FeatureAnalysisModelProtos";

message FeatureAnalysisModel {
  enum FeatureType {
    UNKNOWN_FEATURETYPE = 0;
    NUMERICAL = 1;
    CATEGORICAL = 2;
    CROSS = 3;
    VECTOR = 4;
    MULTI_CATEGORICAL = 5;
  }

  enum ValueType {
    UNKNOWN_VALUETYPE = 0;
    LONG = 1;
    INT = 2;
    DOUBLE = 3;
    FLOAT = 4;
    STRING = 5;
    CROSS_VALUE = 6;
    DOUBLE_VEC = 7;
    FLOAT_VEC = 8;  // repeated float
  }

  enum EncoderType {
    UNKNOWN_ENCODERTYPE = 0;
    RAWVALUE = 1;
    INDEX = 2;
    SEGMENT = 3;
    MULTIINDEX = 4;
  }
  
  message BasicInfo {
    optional string feature_name = 1;
    optional string data_type = 2;
    optional uint64 distinct_cnt = 3;  // 去重值的数量
    optional double missing_ratio = 4;  // 缺失值的比例
    optional FeatureType suggested_feature_type = 5;
  }
  
  message BucketInfo {  // For Categorical features
    optional string feature_name = 1;
    map<string, uint64> bucket_map = 2;  // key field cannot be any
  }

  message ConfigInfo {
    optional string feature_name = 1;
    optional string default_value = 2;
    optional double numerical_outlier_lower_bound = 3;
    optional double numerical_outlier_upper_bound = 4;
    optional double numerical_outlier_lower_ratio = 5;
    optional double numerical_outlier_upper_ratio = 6;
    optional double categorical_outlier_low_ratio = 7;
    optional string delimiter = 8;
  }
  
  message SegmentInfo {  // For Numerical Features
    optional string feature_name = 1;
    repeated double segment_list = 2 [packed=true];
    repeated double info_value_list = 3 [packed=true];
  }

  optional string feature_name = 1;
  optional uint64 feature_id = 2;
  optional FeatureType feature_type = 3;
  optional ConfigInfo config_info = 4;
  optional BasicInfo basic_info = 5;
  optional BucketInfo bucket_info = 6;
  optional SegmentInfo segment_info = 7;
  optional EncoderType encoder_type = 8;
  optional uint64 offset = 9;
}

message FeatureAnalysisModelVec {
  repeated FeatureAnalysisModel feature_analysis_model = 1;
  optional uint64 total_size = 2;
}
